File: src/me/newyith/util/Point.java
package me.newyith.util;

import me.newyith.memory.Memorable;
import me.newyith.memory.Memory;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;

public class Point implements Memorable {
	public World world;
	public double x, y, z;

	public void saveTo(Memory m) {
		//*
		String s = "";
		s += world.getName();
		s += ",";
		s += (int) x;
		s += ",";
		s += (int) y;
		s += ",";
		s += (int) z;
		m.save("s", s);
		/*/
		m.save("worldName", world.getName());
		m.save("x", (int) x);
		m.save("y", (int) y);
		m.save("z", (int) z);
		//*/
	}

	public static Point loadFrom(Memory m) {
		//*
		String s = m.loadString("s");
		String[] data = s.split(",");
		World world = Bukkit.getWorld(data[0]);
		int x = Integer.valueOf(data[1]);
		int y = Integer.valueOf(data[2]);
		int z = Integer.valueOf(data[3]);
		return new Point(world, x, y, z);
		/*/
		String worldName = m.loadString("worldName");
		World world = Bukkit.getWorld(worldName);
		int x = m.loadInt("x");
		int y = m.loadInt("y");
		int z = m.loadInt("z");
		return new Point(world, x, y, z);
		//*/
	}

	//------------------------------------------------------------------------------------------------------------------

	public Point(Point p) {
		this.world = p.world;
		this.x = p.x;
		this.y = p.y;
		this.z = p.z;
	}

	public Point(Location loc) {
		this.world = loc.getWorld();
		this.x = loc.getBlockX();
		this.y = loc.getBlockY();
		this.z = loc.getBlockZ();
	}

	public Point(World world, double x, double y, double z) {
		this.world = world;
		this.x = x;
		this.y = y;
		this.z = z;
	}

	public World getWorld() {
		return world;
	}

	public Block getBlock() {
		return world.getBlockAt((int)x, (int)y, (int)z);
	}

	public boolean matches(Material material) {
		return this.getBlock().getType() == material;
	}

	public Location toLocation() {
		return new Location(this.world, this.x, this.y, this.z);
	}

	public Point difference(Point p) {
		Point d = new Point(this);
		d.x -= p.x;
		d.y -= p.y;
		d.z -= p.z;
		return d;
	}

	public void add(Point p) {
		if(p != null && p.getWorld() == this.getWorld()) {
			this.x += p.x;
			this.y += p.y;
			this.z += p.z;
		} else {
			throw new IllegalArgumentException("Cannot add Locations of differing worlds");
		}
	}

	public Point add(double x, double y, double z) {
		this.x += x;
		this.y += y;
		this.z += z;
		return this;
	}

	public boolean is(Material mat) {
		return getBlock().getType() == mat;
	}

	@Override
	public String toString() {
		int x = (int)this.x;
		int y = (int)this.y;
		int z = (int)this.z;

		return x + ", " + y + ", " + z;
	}

	public String toStringDoubles() {
		return x + ", " + y + ", " + z;
	}

	@Override
	public boolean equals(Object o) {
		if (o == this) {
			return true;
		}

        if (o instanceof Point) {
			Point p = (Point)o;

			boolean equal = true;
			equal = equal && (int)this.x == (int)p.x;
			equal = equal && (int)this.y == (int)p.y;
			equal = equal && (int)this.z == (int)p.z;

			return equal;
		} else {
			return false;
		}
	}

	@Override
	public int hashCode() {
		int hash = (int)x;
		hash = 49999 * hash + (int)y;
		hash = 49999 * hash + (int)z;
		return hash;
	}
}

File: src/main/java/me/newyith/util/Point.kt
package me.newyith.util

import com.fasterxml.jackson.annotation.JsonProperty
import org.bukkit.Chunk
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.util.Vector

open class Point (
	@JsonProperty("x") val x: Double,
	@JsonProperty("y") val y: Double,
	@JsonProperty("z") val z: Double
) {
	constructor(x: Int, y: Int, z: Int) : this(x.toDouble(), y.toDouble(), z.toDouble())
	constructor(loc: Location) : this(loc.x, loc.y, loc.z)
	constructor(v: Vector) : this(v.x, v.y, v.z)
	constructor(b: Block) : this(b.location)

	// --- //

	val xInt
		get() = x.toInt()
	val yInt
		get() = y.toInt()
	val zInt
		get() = z.toInt()

	fun getChunk(world: World): Chunk {
		return toLocation(world).chunk
	}

	fun toVector(): Vector {
		return Vector(x, y, z)
	}

	fun toLocation(world: World): Location {
		return Location(world, x, y, z)
	}

	fun difference(p: Point): Point {
		return Point(
			x - p.x,
			y - p.y,
			z - p.z
		)
	}

	fun distance(p: Point): Double {
		val v1 = this.toVector()
		val v2 = p.toVector()
		return v1.distance(v2)
	}

	fun add(p: Point): Point {
		return Point(
			x + p.x,
			y + p.y,
			z + p.z
		)
	}

	fun add(xAdd: Int, yAdd: Int, zAdd: Int): Point {
		return Point(
			x + xAdd,
			y + yAdd,
			z + zAdd
		)
	}

	fun add(xAdd: Double, yAdd: Double, zAdd: Double): Point {
		return Point(
			x + xAdd,
			y + yAdd,
			z + zAdd
		)
	}

	fun isType(mat: Material, world: World): Boolean {
		return getBlock(world).type == mat
	}

	fun getBlock(world: World): Block {
		return world.getBlockAt(xInt, yInt, zInt)
	}

	fun getType(world: World): Material {
		return getBlock(world).type
	}

	fun getAdjacent6(): Set<Point> {
		val points = HashSet<Point>()

		points.add(this.add(1, 0, 0))
		points.add(this.add(-1, 0, 0))
		points.add(this.add(0, 1, 0))
		points.add(this.add(0, -1, 0))
		points.add(this.add(0, 0, 1))
		points.add(this.add(0, 0, -1))

		return points
	}

	fun getAdjacent26(): Set<Point> {
		val points = HashSet<Point>()
		val range = arrayOf(-1, 0, 1)

		for (x in range) {
			for (y in range) {
				for (z in range) {
					val isCenter = x == 0 && y == 0 && z == 0
					if (!isCenter) points.add(this.add(x, y, z))
				}
			}
		}

		return points
	}

	// - Overrides - //

	override fun toString(): String {
		val s = StringBuilder()
		s.append(xInt)
		s.append(", ")
		s.append(yInt)
		s.append(", ")
		s.append(zInt)
		return s.toString()
	}

	override fun equals(other: Any?): Boolean {
		if (other === this) return true

		if (other is Point) {
			val p = other

			var equal = true
			equal = equal && Math.floor(x) == Math.floor(p.x)
			equal = equal && Math.floor(y) == Math.floor(p.y)
			equal = equal && Math.floor(z) == Math.floor(p.z)

			return equal
		} else {
			return false
		}
	}

	override fun hashCode(): Int {
		var hash = Math.floor(x).toInt()
		hash = 49999 * hash + Math.floor(y).toInt()
		hash = 49999 * hash + Math.floor(z).toInt()
		return hash
	}
}
